File: microservice-architecture/retry-pattern/ts/test/test_retry.ts
import { retry, ScriptedOperation } from "../src/retry";

function assertEqual<T>(expected: T, actual: T, message: string): void {
  if (expected !== actual) throw new Error(`${message} expected=${expected} actual=${actual}`);
}

const op1 = new ScriptedOperation(2);
const result1 = retry(3, () => op1.call());
assertEqual(true, result1.ok, "retry should succeed");
assertEqual(3, result1.attempts, "retry attempts");

const op2 = new ScriptedOperation(5);
const result2 = retry(3, () => op2.call());
assertEqual(false, result2.ok, "retry should fail");
assertEqual(3, result2.attempts, "retry max attempts");

console.log("retry-pattern(ts) tests passed");

File: microservice-architecture/retry-pattern/python/test/test_retry.py
import os
import sys
import unittest

sys.path.append(os.getcwd())

from src.retry import ScriptedOperation, retry  # noqa: E402


class RetryPatternTest(unittest.TestCase):
    def test_retry_succeeds(self) -> None:
        ok, attempts = retry(3, ScriptedOperation(2).call)
        self.assertTrue(ok)
        self.assertEqual(3, attempts)

    def test_retry_fails_after_max_attempts(self) -> None:
        ok, attempts = retry(3, ScriptedOperation(5).call)
        self.assertFalse(ok)
        self.assertEqual(3, attempts)


if __name__ == "__main__":
    unittest.main()
